tachyon/zk/plonk/vanishing/circuit_polynomial_builder.h

```diff
@@ -131,6 +131,51 @@ class CircuitPolynomialBuilder {
       PermutationEvaluator<Evals>& permutation_evaluator,
       lookup::Evaluator<LS::kType, Evals>& lookup_evaluator,
       shuffle::Evaluator<EvalsOrExtendedEvals>& shuffle_evaluator) {
+    if constexpr (Vendor == halo2::Vendor::kPSE) {
+      return BuildExtendedCircuitColumnPSE(custom_gate_evaluator,
+                                           permutation_evaluator,
+                                           lookup_evaluator, shuffle_evaluator);
+    } else {
+      return BuildExtendedCircuitColumnScroll(
+          custom_gate_evaluator, permutation_evaluator, lookup_evaluator,
+          shuffle_evaluator);
+    }
+  }
+
+ private:
+  friend class CustomGateEvaluator<EvalsOrExtendedEvals>;
+  friend class PermutationEvaluator<EvalsOrExtendedEvals>;
+  friend class lookup::halo2::Evaluator<EvalsOrExtendedEvals>;
+  friend class lookup::log_derivative_halo2::Evaluator<EvalsOrExtendedEvals>;
+  friend class shuffle::Evaluator<EvalsOrExtendedEvals>;
+
+  EvaluationInput<EvalsOrExtendedEvals> ExtractEvaluationInput(
+      std ::vector<F>&& intermediates, std::vector<int32_t>&& rotations) {
+    return EvaluationInput<EvalsOrExtendedEvals>(std::move(intermediates),
+                                                 std::move(rotations), table_,
+                                                 theta_, beta_, gamma_, y_, n_);
+  }
+
+  void UpdateLPolyCosets() {
+    l_first_ = coset_domain_->FFT(proving_key_.l_first());
+    l_last_ = coset_domain_->FFT(proving_key_.l_last());
+    l_active_row_ = coset_domain_->FFT(proving_key_.l_active_row());
+  }
+
+  ExtendedEvals BuildExtendedCircuitColumnPSE(
+      CustomGateEvaluator<Evals>& custom_gate_evaluator,
+      PermutationEvaluator<Evals>& permutation_evaluator,
+      lookup::Evaluator<LS::kType, Evals>& lookup_evaluator,
+      shuffle::Evaluator<EvalsOrExtendedEvals>& shuffle_evaluator) {
+    NOTREACHED();
+    return {};
+  }
```

Comment:
Let's just squash with the last commit and change the commit message.

Reply:
This change was necessary because the types of `l_first_`, `l_last_`, and `l_active_row_` are `ExtendedEvals` in the PSE version. If `UpdateLPolyCosets()` is called in the PSE version, it would cause a compile error.